File: vmp-ts/src/type_registry.ts
/**
 * Type registry for ZData encoding/decoding.
 *
 * Provides a global registry that can be extended by users to add custom types.
 * This module matches the Python vmp-py TypeRegistry architecture.
 */

import type { ZDataDict, TypeEncoder, TypeDecoder, TypeChecker } from './types.js';

/**
 * Registry entry for type checkers
 */
interface TypeCheckerEntry {
  checker: TypeChecker;
  typeName: string;
  encoder: TypeEncoder;
}

/**
 * Registry for custom type encoders and decoders.
 *
 * This registry can be extended by users to add support for custom types.
 * The registry supports three ways of registering types:
 * 1. By exact type class/constructor
 * 2. By custom type checker function
 * 3. Both (recommended for reliability)
 */
export class TypeRegistry {
  private encoders = new Map<Function, { typeName: string; encoder: TypeEncoder }>();
  private decoders = new Map<string, TypeDecoder>();
  private typeCheckers: TypeCheckerEntry[] = [];

  /**
   * Register a type for encoding/decoding.
   *
   * @param typeName - Unique identifier for this type (e.g., "numpy.ndarray")
   * @param encoder - Function that encodes data to ZDataDict
   * @param decoder - Function that decodes ZDataDict back to original type
   * @param typeClass - Optional type class/constructor for direct type checking
   * @param typeChecker - Optional function for custom type checking
   *
   * @example
   * ```typescript
   * class Point {
   *   constructor(public x: number, public y: number) {}
   * }
   *
   * function encodePoint(p: Point): ZDataDict {
   *   return {
   *     ztype: "custom.Point",
   *     b: new TextEncoder().encode(`${p.x},${p.y}`)
   *   };
   * }
   *
   * function decodePoint(z: ZDataDict): Point {
   *   const [x, y] = new TextDecoder().decode(z.b).split(",").map(Number);
   *   return new Point(x, y);
   * }
   *
   * TYPE_REGISTRY.register(
   *   "custom.Point",
   *   encodePoint,
   *   decodePoint,
   *   Point  // type class
   * );
   * ```
   */
  register(
    typeName: string,
    encoder: TypeEncoder,
    decoder: TypeDecoder,
    typeClass?: Function,
    typeChecker?: TypeChecker
  ): void {
    // Register decoder
    this.decoders.set(typeName, decoder);

    // Register encoder by type class
    if (typeClass !== undefined) {
      this.encoders.set(typeClass, { typeName, encoder });
    }

    // Register encoder by type checker
    if (typeChecker !== undefined) {
      this.typeCheckers.push({ checker: typeChecker, typeName, encoder });
    }
  }

  /**
   * Encode data using registered encoders.
   *
   * Returns the encoded ZDataDict if a matching encoder is found,
   * otherwise returns the data unchanged.
   *
   * @param data - Data to encode
   * @returns ZDataDict or original data
   */
  encode(data: unknown): unknown {
    // Check by exact type/constructor
    if (data !== null && typeof data === 'object') {
      const constructor = (data as object).constructor;
      if (constructor && this.encoders.has(constructor)) {
        const entry = this.encoders.get(constructor)!;
        return entry.encoder(data);
      }
    }

    // Check using custom type checkers
    for (const { checker, encoder } of this.typeCheckers) {
      if (checker(data)) {
        return encoder(data);
      }
    }

    // No encoder found, return as-is
    return data;
  }

  /**
   * Decode ZData using registered decoders.
   *
   * Returns the decoded object if zdata is a valid ZData dict,
   * otherwise returns the data unchanged.
   *
   * @param zdata - ZData dictionary or plain data
   * @returns Decoded object or original data
   * @throws TypeError if zdata has an unknown ztype
   */
  decode(zdata: unknown): unknown {
    if (!this.isZData(zdata)) {
      return zdata;
    }

    const zdataDict = zdata as ZDataDict;
    const ztype = zdataDict.ztype;

    if (!this.decoders.has(ztype)) {
      throw new TypeError(`Unknown ZData type: ${ztype}`);
    }

    const decoder = this.decoders.get(ztype)!;
    return decoder(zdataDict);
  }

  /**
   * Check if data is a ZData encoded object.
   *
   * @param data - Data to check
   * @returns True if data is a ZData dictionary
   */
  isZData(data: unknown): data is ZDataDict {
    return (
      typeof data === 'object' &&
      data !== null &&
      'ztype' in data &&
      typeof (data as ZDataDict).ztype === 'string'
    );
  }

  /**
   * Get the ztype of a ZData object.
   *
   * @param data - Data to check
   * @returns The ztype string if data is ZData, undefined otherwise
   */
  getZType(data: unknown): string | undefined {
    if (this.isZData(data)) {
      return (data as ZDataDict).ztype;
    }
    return undefined;
  }

  /**
   * List all registered type names.
   *
   * @returns Array of registered ztype names
   */
  listRegisteredTypes(): string[] {
    return Array.from(this.decoders.keys());
  }
}

/**
 * Global type registry - users can extend this
 */
export const TYPE_REGISTRY = new TypeRegistry();

File: vmp-py/src/vuer_vrpc/type_registry.py
"""
Type registry for ZData encoding/decoding.

Provides a global registry that can be extended by users to add custom types.
This module can be imported and extended in user code or third-party libraries.
"""

from typing import Any, Callable, Dict, Optional, Protocol
from typing_extensions import TypedDict


class ZDataDict(TypedDict, total=False):
    """Type definition for ZData encoded objects."""
    ztype: str
    b: bytes
    dtype: Optional[str]
    shape: Optional[tuple]


class TypeEncoder(Protocol):
    """Protocol for type encoders."""
    def __call__(self, data: Any) -> ZDataDict:
        """Encode data to ZData format."""
        ...


class TypeDecoder(Protocol):
    """Protocol for type decoders."""
    def __call__(self, zdata: ZDataDict) -> Any:
        """Decode ZData back to original type."""
        ...


class TypeRegistry:
    """
    Registry for custom type encoders and decoders.

    This registry can be extended by users to add support for custom types.
    The registry supports three ways of registering types:
    1. By exact type class
    2. By custom type checker function
    3. By import path (for lazy registration)
    """

    def __init__(self):
        self._encoders: Dict[type, tuple[str, TypeEncoder]] = {}
        self._decoders: Dict[str, TypeDecoder] = {}
        self._type_checkers: list[tuple[Callable[[Any], bool], str, TypeEncoder]] = []

    def register(
        self,
        type_name: str,
        encoder: TypeEncoder,
        decoder: TypeDecoder,
        type_class: Optional[type] = None,
        type_checker: Optional[Callable[[Any], bool]] = None
    ):
        """
        Register a type for encoding/decoding.

        Args:
            type_name: Unique identifier for this type (e.g., "numpy.ndarray")
            encoder: Function that encodes data to ZDataDict
            decoder: Function that decodes ZDataDict back to original type
            type_class: Optional type class for direct type checking
            type_checker: Optional function for custom type checking

        Example:
            >>> def encode_point(p):
            ...     return {"ztype": "custom.Point", "b": f"{p.x},{p.y}".encode()}
            >>> def decode_point(z):
            ...     x, y = map(float, z["b"].decode().split(","))
            ...     return Point(x, y)
            >>> TYPE_REGISTRY.register(
            ...     "custom.Point",
            ...     encode_point,
            ...     decode_point,
            ...     type_class=Point
            ... )
        """
        # Register decoder
        self._decoders[type_name] = decoder

        # Register encoder by type class
        if type_class is not None:
            self._encoders[type_class] = (type_name, encoder)

        # Register encoder by type checker
        if type_checker is not None:
            self._type_checkers.append((type_checker, type_name, encoder))

    def encode(self, data: Any) -> Any:
        """
        Encode data using registered encoders.

        Returns the encoded ZDataDict if a matching encoder is found,
        otherwise returns the data unchanged.
        """
        # Check by exact type
        data_type = type(data)
        if data_type in self._encoders:
            type_name, encoder = self._encoders[data_type]
            return encoder(data)

        # Check using custom type checkers
        for checker, type_name, encoder in self._type_checkers:
            if checker(data):
                return encoder(data)

        # No encoder found, return as-is
        return data

    def decode(self, zdata: Any) -> Any:
        """
        Decode ZData using registered decoders.

        Returns the decoded object if zdata is a valid ZData dict,
        otherwise returns the data unchanged.
        """
        if not isinstance(zdata, dict) or "ztype" not in zdata:
            return zdata

        ztype = zdata["ztype"]
        if ztype not in self._decoders:
            raise TypeError(f"Unknown ZData type: {ztype}")

        decoder = self._decoders[ztype]
        return decoder(zdata)

    def is_zdata(self, data: Any) -> bool:
        """Check if data is a ZData encoded object."""
        return isinstance(data, dict) and "ztype" in data

    def get_ztype(self, data: Any) -> Optional[str]:
        """Get the ztype of a ZData object."""
        if self.is_zdata(data):
            return data["ztype"]
        return None

    def list_registered_types(self) -> list[str]:
        """List all registered type names."""
        return list(self._decoders.keys())


# Global type registry - users can extend this
TYPE_REGISTRY = TypeRegistry()
